core/sr-io/without_std.rs

```diff
@@ -43,41 +51,214 @@ pub extern fn oom(_: ::core::alloc::Layout) -> ! {
 	static OOM_MSG: &str = "Runtime memory exhausted. Aborting";
 
 	unsafe {
-		ext_print_utf8(OOM_MSG.as_ptr(), OOM_MSG.len() as u32);
+		extern_functions_host_impl::ext_print_utf8(OOM_MSG.as_ptr(), OOM_MSG.len() as u32);
 		intrinsics::abort();
 	}
 }
 
+/// A function which implementation can be exchanged.
+///
+/// Internally this works by swapping function pointers.
+pub struct ExchangeableFunction<T>(T);
+
+impl<T> ExchangeableFunction<T> {
+	/// Replace the implementation with `new_impl`.
+	pub fn replace_implementation(self: &'static mut Self, new_impl: T)  -> RestoreImplementation<T> {
+		let old = mem::replace(&mut self.0, new_impl);
+
+		RestoreImplementation(self, Some(old))
+	}
+
+	/// Set the given implementation.
+	fn set_implementation(&mut self, new_impl: T) {
+		self.0 = new_impl;
+	}
+}
+
+impl<T> Deref for ExchangeableFunction<T> {
+	type Target = T;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+/// Restores a function implementation on drop.
+pub struct RestoreImplementation<T: 'static>(&'static mut ExchangeableFunction<T>, Option<T>);
```

Comment:
why does this require `'static` anyway? can't this be `'a`?

Reply:
```
   |                                  help: consider adding an explicit lifetime bound `T: 'static`...
   |
note: ...so that the reference type `&'static mut ExchangeableFunction<T>` does not outlive the data it points at
```